src/java/org/apache/cassandra/cql3/restrictions/ClusteringColumnRestrictions.java

```diff
@@ -92,24 +92,16 @@ public ClusteringColumnRestrictions mergeWith(Restriction restriction, @Nullable
         return new ClusteringColumnRestrictions(this.comparator, newRestrictionSet, allowFiltering);
     }
 
-    private boolean hasMultiColumnSlice()
-    {
-        for (SingleRestriction restriction : restrictions)
-        {
-            if (restriction.isMultiColumn() && restriction.isSlice())
-                return true;
-        }
-        return false;
-    }
-
     public NavigableSet<Clustering<?>> valuesAsClustering(QueryOptions options, ClientState state) throws InvalidRequestException
     {
-        MultiCBuilder builder = MultiCBuilder.create(comparator, hasIN());
+        MultiCBuilder builder = new MultiCBuilder(comparator);
         for (SingleRestriction r : restrictions)
         {
-            r.appendTo(builder, options);
+            List<ClusteringElements> values = r.values(options);
+            builder.extend(values);
 
-            if (hasIN() && Guardrails.inSelectCartesianProduct.enabled(state))
+            // If values is greater than 1 we know that the restriction is an IN
+            if (values.size() > 1 && Guardrails.inSelectCartesianProduct.enabled(state))
```

Comment:
Would `hasIN()` be incorrect though? Seems more explicit/future-proof, but there's probably something I'm missing...

Reply:
One goal of the patch was to remove the concept of operators when you look at a set of restrictions. There is no `isIN` any more in this class.